crypto/err/err_mark.c

```diff
@@ -26,6 +26,19 @@ int ERR_set_mark(void)
     return 1;
 }
 
+int ERR_pop(void)
+{
+    ERR_STATE *es;
+
+    es = ossl_err_get_state_int();
+    if (es == NULL || es->bottom == es->top)
+        return 0;
+
+    err_clear(es, es->top, 0);
+    es->top = es->top > 0 ? es->top - 1 : ERR_NUM_ERRORS - 1;
+    return 1;
+}
+
```

Comment:
This seems to make the assumption that the caller knows how many errors to expect in a given call chain.  While that makes sense in your upcoming use cases, is it prone to inadvertent failure if future changes increase the number of errors pushed to the stack? I.e. will a single ERR_pop no longer be sufficient?

Reply:
One general principle we've always followed is that libssl never depends on non-public APIs in libcrypto. This means when libssl interacts with libcrypto, we have to make that a public API &mdash; or build the same source file twice, once in libcrypto, once in libssl (e.g. `ossl_crypto_mutex`). We can't do that for ERR either since it relies on thread locals which live in libcrypto.

TBH I'm deeply surprised we didn't have this API already. It seems rather important. Anyway, we can add public APIs in feature releases, and I think this is a desirable addition.